File: web/src/grid.js
/**
 * Block Blast - Grid Logic
 * Manages the game board and line detection
 */
import { pickShapeByCategoryWeights } from './shapes.js';

export class Grid {
    constructor(size = 8) {
        this.size = size;
        this.cells = this.createEmptyGrid();
    }

    createEmptyGrid() {
        const grid = [];
        for (let i = 0; i < this.size; i++) {
            grid[i] = [];
            for (let j = 0; j < this.size; j++) {
                grid[i][j] = null;
            }
        }
        return grid;
    }

    clear() {
        this.cells = this.createEmptyGrid();
    }

    clone() {
        const grid = new Grid(this.size);
        for (let y = 0; y < this.size; y++) {
            for (let x = 0; x < this.size; x++) {
                grid.cells[y][x] = this.cells[y][x];
            }
        }
        return grid;
    }

    canPlace(shape, gx, gy) {
        for (let y = 0; y < shape.length; y++) {
            for (let x = 0; x < shape[y].length; x++) {
                if (shape[y][x]) {
                    const nx = gx + x;
                    const ny = gy + y;
                    if (nx < 0 || nx >= this.size || ny < 0 || ny >= this.size) {
                        return false;
                    }
                    if (this.cells[ny][nx] !== null) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    /**
     * 形状以 (gx,gy) 落子后，各格中心在棋盘上的平均坐标（格坐标系，可为小数）
     */
    static shapeCenterOnBoard(shape, gx, gy) {
        let sxa = 0;
        let sya = 0;
        let n = 0;
        for (let y = 0; y < shape.length; y++) {
            for (let x = 0; x < shape[y].length; x++) {
                if (shape[y][x]) {
                    sxa += gx + x + 0.5;
                    sya += gy + y + 0.5;
                    n++;
                }
            }
        }
        if (!n) {
            return { x: gx + 0.5, y: gy + 0.5 };
        }
        return { x: sxa / n, y: sya / n };
    }

    /**
     * 仅在锚点 (anchorX, anchorY) 附近（切比雪夫半径 radius）的合法落点中，
     * 选形状重心离 (aimCx, aimCy) 最近的一处——只纠偏，不按消除数选全盘最优。
     */
    pickNearestLocalPlacement(shape, aimCx, aimCy, anchorX, anchorY, radius) {
        let best = null;
        let bestDist = Infinity;

        for (let dy = -radius; dy <= radius; dy++) {
            for (let dx = -radius; dx <= radius; dx++) {
                const gx = anchorX + dx;
                const gy = anchorY + dy;
                if (!this.canPlace(shape, gx, gy)) {
                    continue;
                }

                const com = Grid.shapeCenterOnBoard(shape, gx, gy);
                const distSq = (com.x - aimCx) ** 2 + (com.y - aimCy) ** 2;

                if (best === null) {
                    best = { x: gx, y: gy };
                    bestDist = distSq;
                    continue;
                }

                if (distSq < bestDist - 1e-8) {
                    best = { x: gx, y: gy };
                    bestDist = distSq;
                } else if (Math.abs(distSq - bestDist) <= 1e-8) {
                    if (gy < best.y || (gy === best.y && gx < best.x)) {
                        best = { x: gx, y: gy };
                    }
                }
            }
        }
        return best;
    }

    place(shape, colorIdx, gx, gy) {
        for (let y = 0; y < shape.length; y++) {
            for (let x = 0; x < shape[y].length; x++) {
                if (shape[y][x]) {
                    this.cells[gy + y][gx + x] = colorIdx;
                }
            }
        }
    }

    checkLines() {
        const fullRows = [];
        const fullCols = [];
        const clearedCells = [];
        const clearedSet = {};

        for (let y = 0; y < this.size; y++) {
            if (this.cells[y].every(c => c !== null)) {
                fullRows.push(y);
            }
        }

        for (let x = 0; x < this.size; x++) {
            let colFull = true;
            for (let y = 0; y < this.size; y++) {
                if (this.cells[y][x] === null) {
                    colFull = false;
                    break;
                }
            }
            if (colFull) {
                fullCols.push(x);
            }
        }

        for (const y of fullRows) {
            for (let x = 0; x < this.size; x++) {
                const key = x + ',' + y;
                if (!clearedSet[key]) {
                    clearedSet[key] = true;
                    clearedCells.push({ x, y, color: this.cells[y][x] });
                }
            }
        }

        for (const x of fullCols) {
            for (let y = 0; y < this.size; y++) {
                const key = x + ',' + y;
                if (!clearedSet[key]) {
                    clearedSet[key] = true;
                    clearedCells.push({ x, y, color: this.cells[y][x] });
                }
            }
        }

        for (const y of fullRows) {
            for (let x = 0; x < this.size; x++) {
                this.cells[y][x] = null;
            }
        }

        for (const x of fullCols) {
            for (let y = 0; y < this.size; y++) {
                this.cells[y][x] = null;
            }
        }

        const lines = fullRows.length + fullCols.length;
        return { count: lines, cells: clearedCells };
    }

    hasAnyMove(blocks) {
        for (const block of blocks) {
            if (!block || block.placed) continue;
            for (let y = 0; y < this.size; y++) {
                for (let x = 0; x < this.size; x++) {
                    if (this.canPlace(block.shape, x, y)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    canPlaceAnywhere(shapeData) {
        for (let y = 0; y < this.size; y++) {
            for (let x = 0; x < this.size; x++) {
                if (this.canPlace(shapeData, x, y)) {
                    return true;
                }
            }
        }
        return false;
    }

    /** 返回该形状在当前盘面上的合法放置位数量 */
    countValidPlacements(shapeData) {
        let count = 0;
        for (let y = 0; y < this.size; y++) {
            for (let x = 0; x < this.size; x++) {
                if (this.canPlace(shapeData, x, y)) count++;
            }
        }
        return count;
    }

    wouldClear(shapeData, gx, gy) {
        const temp = this.cells.map(row => [...row]);
        for (let y = 0; y < shapeData.length; y++) {
            for (let x = 0; x < shapeData[y].length; x++) {
                if (shapeData[y][x]) {
                    temp[gy + y][gx + x] = 1;
                }
            }
        }

        for (let y = 0; y < this.size; y++) {
            if (temp[y].every(c => c !== null)) {
                return true;
            }
        }

        for (let x = 0; x < this.size; x++) {
            let colFull = true;
            for (let y = 0; y < this.size; y++) {
                if (temp[y][x] === null) {
                    colFull = false;
                    break;
                }
            }
            if (colFull) return true;
        }

        return false;
    }

    /**
     * 推演：若在此位置放置方块，放置后哪些整行/整列会被填满（与 checkLines 判定一致，不落子）。
     * @param {number[][]} shapeData
     * @param {number} gx
     * @param {number} gy
     * @param {number} colorIdx 与真实落子一致，用于着色推演
     * @returns {{ rows: number[], cols: number[], cells: { x: number, y: number, color: number }[] } | null} 不可放置时为 null
     */
    previewClearOutcome(shapeData, gx, gy, colorIdx) {
        if (!this.canPlace(shapeData, gx, gy)) {
            return null;
        }
        const temp = this.cells.map((row) => [...row]);
        for (let y = 0; y < shapeData.length; y++) {
            for (let x = 0; x < shapeData[y].length; x++) {
                if (shapeData[y][x]) {
                    temp[gy + y][gx + x] = colorIdx;
                }
            }
        }

        const fullRows = [];
        const fullCols = [];
        for (let y = 0; y < this.size; y++) {
            if (temp[y].every((c) => c !== null)) {
                fullRows.push(y);
            }
        }
        for (let x = 0; x < this.size; x++) {
            let colFull = true;
            for (let y = 0; y < this.size; y++) {
                if (temp[y][x] === null) {
                    colFull = false;
                    break;
                }
            }
            if (colFull) {
                fullCols.push(x);
            }
        }

        const clearedSet = {};
        const cells = [];
        for (const y of fullRows) {
            for (let x = 0; x < this.size; x++) {
                const key = `${x},${y}`;
                if (!clearedSet[key]) {
                    clearedSet[key] = true;
                    cells.push({ x, y, color: temp[y][x] });
                }
            }
        }
        for (const x of fullCols) {
            for (let y = 0; y < this.size; y++) {
                const key = `${x},${y}`;
                if (!clearedSet[key]) {
                    clearedSet[key] = true;
                    cells.push({ x, y, color: temp[y][x] });
                }
            }
        }

        return { rows: fullRows, cols: fullCols, cells };
    }

    findGapPositions() {
        const gaps = [];

        for (let y = 0; y < this.size; y++) {
            let empty = 0;
            const positions = [];
            for (let x = 0; x < this.size; x++) {
                if (this.cells[y][x] === null) {
                    empty++;
                    positions.push({ x, y });
                }
            }
            if (empty >= 1 && empty <= 3) {
                gaps.push({ type: 'row', y, empty, positions });
            }
        }

        for (let x = 0; x < this.size; x++) {
            let empty = 0;
            const positions = [];
            for (let y = 0; y < this.size; y++) {
                if (this.cells[y][x] === null) {
                    empty++;
                    positions.push({ x, y });
                }
            }
            if (empty >= 1 && empty <= 3) {
                gaps.push({ type: 'col', x, empty, positions });
            }
        }

        gaps.sort((a, b) => a.empty - b.empty);
        return gaps;
    }

    countGapFills(shapeData) {
        let fills = 0;
        const gaps = this.findGapPositions();
        for (const gap of gaps) {
            for (const pos of gap.positions) {
                if (this.canPlace(shapeData, pos.x, pos.y)) {
                    fills += (4 - gap.empty);
                    break;
                }
            }
        }
        return fills;
    }

    initBoard(fillRatio, weights) {
        this.clear();
        let placedCells = 0;
        const targetCells = Math.floor(this.size * this.size * fillRatio);

        for (let attempts = 0; attempts < 100 && placedCells < targetCells; attempts++) {
            const shape = pickShapeByCategoryWeights(weights);
            if (!shape) {
                break;
            }
            const x = Math.floor(Math.random() * (this.size - shape.data[0].length + 1));
            const y = Math.floor(Math.random() * (this.size - shape.data.length + 1));

            if (this.canPlace(shape.data, x, y) && !this.wouldClear(shape.data, x, y)) {
                const colorIdx = Math.floor(Math.random() * 8);
                this.place(shape.data, colorIdx, x, y);
                placedCells += shape.data.flat().filter(c => c).length;
            }
        }
    }

    getFillRatio() {
        let filled = 0;
        for (let y = 0; y < this.size; y++) {
            for (let x = 0; x < this.size; x++) {
                if (this.cells[y][x] !== null) filled++;
            }
        }
        return filled / (this.size * this.size);
    }

    toJSON() {
        return {
            size: this.size,
            cells: this.cells.map(row => [...row])
        };
    }

    fromJSON(data) {
        this.size = data.size || 8;
        this.cells = data.cells.map(row => [...row]);
    }
}

File: rl_pytorch/grid.py
"""与 web/src/grid.js 核心规则一致：落子、满行/列消除。"""

from __future__ import annotations

import random
from .shapes_data import pick_random_shape_weighted


class Grid:
    def __init__(self, size: int = 8):
        self.size = size
        self.cells: list[list[int | None]] = self._empty()

    def _empty(self) -> list[list[int | None]]:
        return [[None] * self.size for _ in range(self.size)]

    def clear(self) -> None:
        self.cells = self._empty()

    def clone(self) -> Grid:
        g = Grid(self.size)
        for y in range(self.size):
            for x in range(self.size):
                g.cells[y][x] = self.cells[y][x]
        return g

    def can_place(self, shape: list[list[int]], gx: int, gy: int) -> bool:
        for y, row in enumerate(shape):
            for x, v in enumerate(row):
                if v:
                    nx, ny = gx + x, gy + y
                    if nx < 0 or nx >= self.size or ny < 0 or ny >= self.size:
                        return False
                    if self.cells[ny][nx] is not None:
                        return False
        return True

    def place(self, shape: list[list[int]], color_idx: int, gx: int, gy: int) -> None:
        for y, row in enumerate(shape):
            for x, v in enumerate(row):
                if v:
                    self.cells[gy + y][gx + x] = color_idx

    def check_lines(self, bonus_block_icons: list[str] | None = None) -> dict:
        full_rows: list[int] = []
        full_cols: list[int] = []
        for y in range(self.size):
            if all(c is not None for c in self.cells[y]):
                full_rows.append(y)
        for x in range(self.size):
            if all(self.cells[y][x] is not None for y in range(self.size)):
                full_cols.append(x)

        def line_is_bonus_mono(vals: list[int | None]) -> bool:
            if not vals or vals[0] is None:
                return False
            first = vals[0]
            if bonus_block_icons:
                bi = bonus_block_icons

                def gi(ci: int) -> str:
                    return str(bi[int(ci) % len(bi)])

                icon0 = gi(int(first))
                return all(v is not None and gi(int(v)) == icon0 for v in vals)
            return all(c == first for c in vals)

        bonus_lines: list[dict] = []
        for y in full_rows:
            row = self.cells[y]
            if line_is_bonus_mono(row):
                bonus_lines.append({"type": "row", "idx": y, "color_idx": row[0]})
        for x in full_cols:
            col = [self.cells[yy][x] for yy in range(self.size)]
            if col[0] is None:
                continue
            if line_is_bonus_mono(col):
                bonus_lines.append({"type": "col", "idx": x, "color_idx": col[0]})

        cleared_cells: list[dict] = []
        seen: set[str] = set()
        for y in full_rows:
            for x in range(self.size):
                k = f"{x},{y}"
                if k not in seen:
                    seen.add(k)
                    cleared_cells.append({"x": x, "y": y, "color": self.cells[y][x]})
        for x in full_cols:
            for y in range(self.size):
                k = f"{x},{y}"
                if k not in seen:
                    seen.add(k)
                    cleared_cells.append({"x": x, "y": y, "color": self.cells[y][x]})

        for y in full_rows:
            for x in range(self.size):
                self.cells[y][x] = None
        for x in full_cols:
            for y in range(self.size):
                self.cells[y][x] = None

        count = len(full_rows) + len(full_cols)
        return {"count": count, "cells": cleared_cells, "bonus_lines": bonus_lines}

    def has_any_move(self, blocks: list[dict]) -> bool:
        for b in blocks:
            if not b or b.get("placed"):
                continue
            shape = b["shape"]
            for gy in range(self.size):
                for gx in range(self.size):
                    if self.can_place(shape, gx, gy):
                        return True
        return False

    def can_place_anywhere(self, shape_data: list[list[int]]) -> bool:
        for gy in range(self.size):
            for gx in range(self.size):
                if self.can_place(shape_data, gx, gy):
                    return True
        return False

    def would_clear(self, shape_data: list[list[int]], gx: int, gy: int) -> bool:
        temp = [row[:] for row in self.cells]
        for y, row in enumerate(shape_data):
            for x, v in enumerate(row):
                if v:
                    temp[gy + y][gx + x] = 1
        for y in range(self.size):
            if all(c is not None for c in temp[y]):
                return True
        for x in range(self.size):
            if all(temp[y][x] is not None for y in range(self.size)):
                return True
        return False

    def find_gap_positions(self) -> list[dict]:
        gaps: list[dict] = []
        for y in range(self.size):
            positions = [{"x": x, "y": y} for x in range(self.size) if self.cells[y][x] is None]
            empty = len(positions)
            if 1 <= empty <= 4:
                gaps.append({"type": "row", "y": y, "empty": empty, "positions": positions})
        for x in range(self.size):
            positions = [{"x": x, "y": y} for y in range(self.size) if self.cells[y][x] is None]
            empty = len(positions)
            if 1 <= empty <= 4:
                gaps.append({"type": "col", "x": x, "empty": empty, "positions": positions})
        gaps.sort(key=lambda g: g["empty"])
        return gaps

    def count_gap_fills(self, shape_data: list[list[int]]) -> int:
        fills = 0
        for gap in self.find_gap_positions():
            for pos in gap["positions"]:
                if self.can_place(shape_data, pos["x"], pos["y"]):
                    fills += max(1, 4 - gap["empty"])
                    break
        return fills

    def init_board(self, fill_ratio: float, shape_weights: dict[str, float] | None = None) -> None:
        self.clear()
        placed_cells = 0
        target = int(self.size * self.size * fill_ratio)
        for _ in range(100):
            if placed_cells >= target:
                break
            shape = pick_random_shape_weighted(shape_weights)
            data = shape["data"]
            w, h = len(data[0]), len(data)
            x = random.randint(0, max(0, self.size - w))
            y = random.randint(0, max(0, self.size - h))
            if self.can_place(data, x, y) and not self.would_clear(data, x, y):
                color = random.randint(0, 7)
                self.place(data, color, x, y)
                placed_cells += sum(1 for row in data for c in row if c)
